dali/test/python/test_fw_iterators.py

```diff
@@ -1024,30 +983,61 @@ def __next__(self):
         if i == 2:
             break
 
-def check_external_source_autoreset(Iterator, *args, **kwargs):
-    batch_size = 4
+def check_external_source_autoreset(Iterator, *args, to_np=None, **kwargs):
+    max_batch_size = 4
     iter_limit = 4
     runs = 3
     test_data_shape = [2, 3, 4]
     i = 0
+    dataset = [[[np.random.randint(0, 255, size = test_data_shape, dtype = np.uint8) for _ in range(max_batch_size)]] for _ in range(iter_limit)]
     def get_data():
         nonlocal i
         if i == iter_limit:
             i = 0
             raise StopIteration
-        out = [[np.random.randint(0, 255, size = test_data_shape, dtype = np.uint8) for _ in range(batch_size)]]
+        out = dataset[i]
         i += 1
         return out
 
-    pipe = Pipeline(batch_size = batch_size, num_threads = 1, device_id = 0)
+    pipe = Pipeline(batch_size = max_batch_size, num_threads = 1, device_id = 0)
     with pipe:
         outs = fn.external_source(source = get_data, num_outputs=1)
     pipe.set_outputs(*outs)
 
     it = Iterator([pipe], *args, auto_reset=True, **kwargs)
     counter = 0
     for _ in range(runs):
-        for __ in enumerate(it):
+        for j, data in enumerate(it):
+            assert (to_np(data[0]) == np.concatenate(dataset[j])).all()
+            counter += 1
+    assert counter == iter_limit * runs
+
+def check_external_source_variable_size(Iterator, *args, iter_size=-1, to_np=None, **kwargs):
+    max_batch_size = 1
+    iter_limit = 4
+    runs = 3
+    test_data_shape = [2, 3, 4]
+    i = 0
+    dataset = [[[np.random.randint(0, 255, size = test_data_shape, dtype = np.uint8) for _ in range(random.randint(1, max_batch_size))]] for _ in range(iter_limit)]
+    def get_data():
+        nonlocal i
+        if i == iter_limit:
+            i = 0
+            raise StopIteration
+        out = dataset[i]
+        i += 1
+        return out
+
+    pipe = Pipeline(batch_size = max_batch_size, num_threads = 1, device_id = 0)
+    with pipe:
+        outs = fn.external_source(source = get_data, num_outputs=1)
+    pipe.set_outputs(*outs)
```

Comment:
Again, I don't really now precisely how this should work, if the `set_outputs` should be in the scope of `with` or not... Just putting attention to that

Reply:
Scope matters for the operators with side effects. I don't think `set_outputs` has any.